### PR TITLE
Eng-6790-Auto-Reconnect-CSV/Kafka/JDBCLoader

### PR DESCRIPTION
**Pull request by sycai**
Add auto-reconnect feature to CSVLoader, KafkaLoader 8 & 10, and JDBCLoader. They pause and wait for the connection when it is lost.

### CODE REVIEW

File: src/frontend/org/voltdb/utils/CSVTupleDataLoader.java [hunk 1]
```diff
@@ -159,25 +160,28 @@ public CSVTupleDataLoader(ClientImpl client, String procName, BulkLoaderErrorHan
...
+                        new PartitionSingleExecuteProcedureCallback(metaData);
+                if (!m_client.callProcedure(cbmt, m_insertProcedure, values)) {
+                    m_log.fatal("Failed to send CSV insert to VoltDB cluster.");
+                    ClientResponse response = new ClientResponseImpl(ClientResponseImpl.SERVER_UNAVAILABLE,
+                            new VoltTable[0], "Failed to call procedure.", 0);
+                    m_errHandler.handleError(metaData, response, "Failed to call procedure.");
+                }
+                // Row inserted successfully, so move on
+                break;
+            } catch (NoConnectionsException ex) {
```
Comment (nshi) [inline src/frontend/org/voltdb/utils/CSVTupleDataLoader.java]: This is a type of IOException, you don't need a separate handler.

File: src/frontend/org/voltdb/utils/CSVTupleDataLoader.java [hunk 2]
```diff
@@ -96,7 +96,8 @@ public void run() {
                     });
                 }
             }
-            else {
+            // Connection lost response is ignored because this data loader has auto-reconnect feature
+            else if (status != ClientResponse.CONNECTION_LOST){
```
Comment (sycai) [inline src/frontend/org/voltdb/utils/CSVTupleDataLoader.java]: `insertRow()` is responsible for connection retry

File: src/frontend/org/voltdb/client/VoltBulkLoader/PerPartitionTable.java [hunk 3]
```diff
@@ -299,4 +299,28 @@ private void loadTable(ProcedureCallback callback, VoltTable toSend) throws Exce
...
+
+        while (true) {
+            try {
+                if (m_isMP) {
+                    m_clientImpl.callProcedure(callback, m_procName, m_tableName, m_upsert, toSend);
+                } else {
+                    Object rpartitionParam = VoltType.valueToBytes(toSend.fetchRow(0).get(
+                            m_partitionedColumnIndex, m_partitionColumnType));
+                    m_clientImpl.callProcedure(callback, m_procName, rpartitionParam, m_tableName, m_upsert, toSend);
+                }
```
Comment (sycai) [inline src/frontend/org/voltdb/client/VoltBulkLoader/PerPartitionTable.java]: Sounds good. I am going to remove `loadTableWithRetry()` and put that retry loop in `loadTable()` instead. It will simplify our code.

File: src/frontend/org/voltdb/utils/CSVTupleDataLoader.java (hunk 2 above)
Comment (sycai) [inline src/frontend/org/voltdb/utils/CSVTupleDataLoader.java]: Yes this is definitely a bug. Will fix it.

File: src/frontend/org/voltdb/client/VoltBulkLoader/PerPartitionTable.java [hunk 4]
```diff
@@ -128,6 +128,39 @@ public void run() {
...
+        final VoltBulkLoaderRow m_reinsertRow;
+
+        ReinsertCallback(VoltBulkLoaderRow row) {
+            m_reinsertRow = row;
+        }
+
+        @Override
+        public void clientCallback(ClientResponse response) throws Exception {
+            if (response.getStatus() == ClientResponse.CONNECTION_LOST) {
+                m_es.execute(new Runnable() {
```
Comment (akhanzode) [inline src/frontend/org/voltdb/client/VoltBulkLoader/PerPartitionTable.java]: PerPartitionTable is under BulkLoader API so wether to resubmit on connection loss depends on client and if its enabled for auto reconnect so per partition table should pick this flag from client that created the bulkloader.
I am debating if we should provide a CLI switch for this going to ask some users. But those who use their own loader using bulkloader should def not get this behavior.

File: src/frontend/org/voltdb/utils/CSVTupleDataLoader.java [hunk 5]
```diff
@@ -159,32 +179,41 @@ public CSVTupleDataLoader(ClientImpl client, String procName, BulkLoaderErrorHan
...
         }
     }
 
     @Override
     public void close() throws InterruptedException, NoConnectionsException
     {
-        m_client.drain();
+        while (m_processedCount.get() != m_totalRowCount.get()) {
+            m_client.drain();
+            Thread.sleep(500);
```
Comment (sycai) [inline src/frontend/org/voltdb/utils/CSVTupleDataLoader.java]: I copied this while loop from VoltBulkLoader.drain() and changed Thread.yield() to Thread.sleep() because I'm more familiar with sleep(). We can always switch to yield() if it is better suited for this scenario.

File: src/frontend/org/voltdb/client/ClientImpl.java [hunk 6]
```diff
@@ -1029,6 +1029,10 @@ public VoltBulkLoader getNewBulkLoader(String tableName, int maxBatchSize, boole
         }
     }
 
+    public boolean isAutoReconnectEnabled() {
```
Comment (akhanzode) [inline src/frontend/org/voltdb/client/ClientImpl.java]: I think this should be also on Client interface.

File: src/frontend/org/voltdb/client/ReconnectStatusListener.java [hunk 7]
```diff
@@ -101,6 +101,10 @@ private void connectToOneServerWithRetry(String hostname, int port) {
             try {
                 m_client.createConnection(hostname, port);
                 LOG.info(String.format("Connected to VoltDB node at %s:%d.", hostname, port));
+                synchronized (m_client) {
```
Comment (akhanzode) [inline src/frontend/org/voltdb/client/ReconnectStatusListener.java]: Why do you need this? Is this the case when cluster is down and cant get connection? Also should this be only when auto connect is enabled?

File: src/frontend/org/voltdb/client/VoltBulkLoader/PerPartitionTable.java [hunk 8]
```diff
@@ -128,8 +132,41 @@ public void run() {
...
+        final VoltBulkLoaderRow m_reinsertRow;
+
+        ReinsertCallback(VoltBulkLoaderRow row) {
+            m_reinsertRow = row;
+        }
+
+        @Override
+        public void clientCallback(ClientResponse response) throws Exception {
+            if (response.getStatus() == ClientResponse.CONNECTION_LOST && m_autoReconnect) {
+                m_reinsertExecutor.execute(new Runnable() {
```
Comment (akhanzode) [inline src/frontend/org/voltdb/client/VoltBulkLoader/PerPartitionTable.java]: Why not use single executor?

File: src/frontend/org/voltdb/client/VoltBulkLoader/PerPartitionTable.java [hunk 9]
```diff
@@ -128,8 +132,41 @@ public void run() {
...
+            m_reinsertRow.m_loader.m_loaderCompletedCnt.incrementAndGet();
+            m_reinsertRow.m_loader.m_outstandingRowCount.decrementAndGet();
+        }
+
+    }
+
     PerPartitionTable(ClientImpl clientImpl, String tableName, int partitionId, boolean isMP,
-            VoltBulkLoader firstLoader, int minBatchTriggerSize, BulkLoaderSuccessCallback successCallback) {
+            VoltBulkLoader firstLoader, int minBatchTriggerSize, BulkLoaderSuccessCallback successCallback,
+            boolean autoReconnect) {
```
Comment (akhanzode) [inline src/frontend/org/voltdb/client/VoltBulkLoader/PerPartitionTable.java]: You dont need to pass this down you can get it from client

File: src/frontend/org/voltdb/client/VoltBulkLoader/PerPartitionTable.java [hunk 10]
```diff
@@ -226,20 +267,8 @@ private void reinsertFailed(List<VoltBulkLoaderRow> rows) throws Exception {
...
-                        row.m_loader.m_notificationCallBack.failureCallback(row.m_rowHandle, row.m_rowData, response);
-                    }
-
-                    row.m_loader.m_loaderCompletedCnt.incrementAndGet();
-                    row.m_loader.m_outstandingRowCount.decrementAndGet();
-                }
-            };
-
+            //one insert at a time callback
+            ProcedureCallback callback = new ReinsertCallback(row);
```
Comment (akhanzode) [inline src/frontend/org/voltdb/client/VoltBulkLoader/PerPartitionTable.java]: This can be done in single ProcedureCallback.

File: src/frontend/org/voltdb/utils/CSVTupleDataLoader.java [hunk 11]
```diff
@@ -127,6 +147,8 @@ public CSVTupleDataLoader(ClientImpl client, String procName, BulkLoaderErrorHan
         m_errHandler = errHandler;
         m_callbackExecutor = callbackExecutor;
         m_successCallback = successCallback;
+        m_reinsertExecutor = CoreUtils.getSingleThreadExecutor(procName);
```
Comment (akhanzode) [inline src/frontend/org/voltdb/utils/CSVTupleDataLoader.java]: You should create this executor only if auto connect is on.

File: src/frontend/org/voltdb/utils/CSVTupleDataLoader.java [hunk 12]
```diff
@@ -159,32 +181,63 @@ public CSVTupleDataLoader(ClientImpl client, String procName, BulkLoaderErrorHan
...
 
     @Override
     public void close() throws InterruptedException, NoConnectionsException
     {
-        m_client.drain();
+        while (m_processedCount.get() != m_totalRowCount.get()) {
+            m_client.drain();
+            Thread.yield();
+        }
+        m_reinsertExecutor.shutdown();
```
Comment (akhanzode) [inline src/frontend/org/voltdb/utils/CSVTupleDataLoader.java]: only shutdown if its present.
